### PR TITLE
Add API to create/access memory

### PR DESCRIPTION
**Pull request by roman-kashitsyn**
This change adds a bunch of functions to manipulate instance memory
and introduces a simple test suite.

### CODE REVIEW

File: src/Wasm/API.hs [hunk 1]
```diff
@@ -414,6 +429,66 @@ wrapFunc store funcType f = do
...
+  withForeignPtr (memoryPtr mem) (\pMem -> NumPages <$> Raw.memorySize pMem)
+
+-- | Returns the size of the provided memory in bytes.
+memoryDataSize :: Memory -> IO Word32
+memoryDataSize mem =
+  withForeignPtr (memoryPtr mem) (\pMem -> Raw.memoryDataSize pMem)
+
+-- | Modify memory by writing the bytestring starting at specified offset.
+--
+-- The portion of the bytestring that spans beyond the memory data size is ignored.
```
Comment (nomeata) [inline src/Wasm/API.hs:454]: Is that idiomatic? Shouldn’t this maybe fail instead?

File: src/Wasm/API.hs [hunk 2]
```diff
@@ -414,6 +429,66 @@ wrapFunc store funcType f = do
...
+writeMemory :: Memory -> Word32 -> ByteString -> IO ()
+writeMemory mem offset bytes =
+  unsafeUseMemory mem $ \pBytes memLen ->
+    UBS.unsafeUseAsCStringLen bytes $ \(pStr, strLen) -> do
+      let n = if memLen < offset then 0 else min (memLen - offset) (fromIntegral strLen)
+      IBS.memcpy (plusPtr pBytes $ fromIntegral offset) (castPtr pStr) (fromIntegral n)
```
Comment (nomeata) [inline src/Wasm/API.hs:460]: Why even do the `memcpy` if the length is `0`? Is `memcpy` defined if the offset is out of bounds, but `n == 0`?

File: src/Wasm/API.hs [hunk 3]
```diff
@@ -414,6 +429,66 @@ wrapFunc store funcType f = do
...
+
+-- | Read the portion of memory specified by offset and size.
+--
+-- The read request is clipped to the memory size, i.e. the read returns the data located in
+-- the intersection of ranges @[offset, offset + size)@ and @[0, memoryDataSize mem)@.
```
Comment (nomeata) [inline src/Wasm/API.hs:465]: Again, I am not sure if it wouldn’t be nicer to guarantee a result of size `size`, and throwing an excpetion when this is an out of bounds read.

File: src/Wasm/API.hs [hunk 4]
```diff
@@ -414,6 +429,66 @@ wrapFunc store funcType f = do
...
+readMemory :: Memory -> Word32 -> Word32 -> IO ByteString
+readMemory mem offset size =
+  unsafeUseMemory mem $ \pBytes memLen -> do
+    let n = fromIntegral $ if memLen < offset then 0 else min size (memLen - offset)
+    IBS.create n (\dst -> IBS.memcpy dst (plusPtr pBytes $ fromIntegral offset) n)
+
+-- | Directly applies an operation to the memory contents.
+--
+-- The operation  accepts a pointer to memory data and the number of
+-- bytes available.
```
Comment (nomeata) [inline src/Wasm/API.hs:475]: How long is that `Ptr` valid? Can I return it from `action`, and use it later (probably not, but better be explicit)

**Review by nomeata — APPROVED** [review]

File: src/Wasm/API.hs (hunk 1 above)
Comment (roman-kashitsyn) [reply src/Wasm/API.hs:461]: Yes, this is the hard choice. I tried to follow the "define errors out of existence" mantra from "A Philosophy of Software Design", but I think most people will find this behaviour surprising. I'll throw `IndexOutOfBounds` instead then.

File: src/Wasm/API.hs (hunk 3 above)
Comment (roman-kashitsyn) [reply src/Wasm/API.hs]: Done.

File: src/Wasm/API.hs (hunk 1 above)
Comment (roman-kashitsyn) [reply src/Wasm/API.hs]: Switched to throwing exceptions.

File: src/Wasm/API.hs (hunk 4 above)
Comment (roman-kashitsyn) [reply src/Wasm/API.hs]: Mentioned the pointer lifetime in the docs.

File: src/Wasm/API.hs (hunk 2 above)
Comment (roman-kashitsyn) [reply src/Wasm/API.hs]: I'm throwing an exception now if offset + strLen > memLen.
I'm pretty sure it's fine to call memcpy with 0 len, especially now given that `plusPtr` always produces a valid pointer.

**Review by nomeata — APPROVED** [review]

**Pull request merged**